flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/BipartiteGraph.java

```diff
@@ -0,0 +1,231 @@
+	private final DataSet<Vertex<KB, VB>> bottomVertices;
+	private final DataSet<BipartiteEdge<KT, KB, EV>> edges;
+
+	private BipartiteGraph(
+			DataSet<Vertex<KT, VT>> topVertices,
+			DataSet<Vertex<KB, VB>> bottomVertices,
+			DataSet<BipartiteEdge<KT, KB, EV>> edges,
+			ExecutionEnvironment context) {
+		this.topVertices = topVertices;
+		this.bottomVertices = bottomVertices;
+		this.edges = edges;
+		this.context = context;
+	}
+
+	/**
+	 * Create bipartite graph from datasets.
+	 *
+	 * @param topVertices  dataset of top vertices in the graph
+	 * @param bottomVertices dataset of bottom vertices in the graph
+	 * @param edges dataset of edges between vertices
+	 * @param context Flink execution context
+	 * @param <KT> the key type of the top vertices
+	 * @param <KB> the key type of the bottom vertices
+	 * @param <VT> the top vertices value type
+	 * @param <VB> the bottom vertices value type
+	 * @param <EV> the edge value type
+	 * @return new bipartite graph created from provided datasets
+	 */
+	public static <KT, KB, VT, VB, EV> BipartiteGraph<KT, KB, VT, VB, EV> fromDataSet(
+			DataSet<Vertex<KT, VT>> topVertices,
+			DataSet<Vertex<KB, VB>> bottomVertices,
+			DataSet<BipartiteEdge<KT, KB, EV>> edges,
+			ExecutionEnvironment context) {
+		return new BipartiteGraph<>(topVertices, bottomVertices, edges, context);
+	}
+
+	/**
+	 * Get dataset with top vertices.
+	 *
+	 * @return dataset with top vertices
+	 */
+	public DataSet<Vertex<KT, VT>> getTopVertices() {
+		return topVertices;
+	}
+
+	/**
+	 * Get dataset with bottom vertices.
+	 *
+	 * @return dataset with bottom vertices
+	 */
+	public DataSet<Vertex<KB, VB>> getBottomVertices() {
+		return bottomVertices;
+	}
+
+	/**
+	 * Get dataset with graph edges.
+	 *
+	 * @return dataset with graph edges
+	 */
+	public DataSet<BipartiteEdge<KT, KB, EV>> getEdges() {
+		return edges;
+	}
+
+	/**
+	 * Convert a bipartite into a graph that contains only top vertices. An edge between two vertices in the new
+	 * graph will exist only if the original bipartite graph contains a bottom vertex they both connected to.
+	 *
+	 * <p>Caller should provide a function that will create an edge between two top vertices. This function will receive
+	 * a collection of all connections between each pair of top vertices. Note that this function will be called twice for
+	 * each pair of connected vertices, so it's up to a caller if one or two edges should be created.
+	 *
+	 * @param edgeFactory function that will be used to create edges in the new graph
+	 * @param <NEV> the edge value type in the new graph
+	 * @return top projection of the bipartite graph
+	 */
+	public <NEV> Graph<KT, VT, NEV> topProjection(
+		final GroupReduceFunction<Tuple2<
+									Tuple2<BipartiteEdge<KT, KB, EV>, BipartiteEdge<KT, KB, EV>>,
+									Vertex<KB, VB>>,
+								Edge<KT, NEV>> edgeFactory) {
+
+		DataSet<Edge<KT, NEV>> newEdges = edges.join(edges)
+			.where(new TopProjectionKeySelector<KT, KB, EV>())
```

Comment:
Using the field index should be faster than a key selector, and allows the optimizer to reuse sorted fields.